ariadne/utils.py

```diff
@@ -0,0 +1,222 @@
+        real_dict[k] = pretrained_dict[needed_key]
+
+    model.load_state_dict(real_dict)
+    model.eval()
+    return model
+
+def find_nearest_hit_no_faiss(ellipses, y, return_numpy=False):
+    centers = ellipses[:, :2]
+    last_station_hits = deepcopy(y)
+    dists = torch.cdist(last_station_hits.float(), centers.float())
+    minimal = last_station_hits[torch.argmin(dists, dim=0)]
+    is_in_ellipse = point_in_ellipse(ellipses, minimal)
+    if return_numpy:
+        minimal = minimal.detach().cpu().numpy()
+        is_in_ellipse  = is_in_ellipse.detach().cpu().numpy()
+    return minimal, is_in_ellipse
+
+def find_nearest_hit(ellipses, last_station_hits):
+    #numpy, numpy -> numpy, numpy
+    index = faiss.IndexFlatL2(2)
+    index.add(last_station_hits.astype('float32'))
+    #ellipses = torch_ellipses.detach().cpu().numpy()
+    centers = ellipses[:,:2]
+    d, i = index.search(np.ascontiguousarray(centers.astype('float32')), 1)
+    x_part = d.flatten() / ellipses[:, 2].flatten()**2
+    y_part = d.flatten() / ellipses[:, 3].flatten()**2
+    left_side = x_part + y_part
+    is_in_ellipse = left_side <= 1
+    return last_station_hits[i.flatten()], is_in_ellipse
+
+def get_diagram_arr_linspace(all_real_hits, found_hits, start, end, num, col):
+    spac = np.linspace(start, end, num=num)
+
+    arr = []
+
+    for i in range(len(spac) - 1):
+        beg = spac[i]
+        end = spac[i + 1]
+        elems_real = all_real_hits[(all_real_hits[col] > beg) & (all_real_hits[col] < end)]
+        elems_pred = found_hits[(found_hits[col] > beg) & (found_hits[col] < end)]
+        if elems_real.empty:
+            arr.append(np.NaN)
+            continue
+        arr.append(len(elems_pred) / len(elems_real))
+
+    return arr, spac[:-1]
+
+def draw_for_col(tracks_real, tracks_pred_true,
+                 col, col_pretty, total_events, n_ticks=150, n_boxes=10, model_name='TrackNETV2.1', style='boxplot'):
+
+    start = tracks_real[tracks_real[col] > -np.inf][col].min()
+    end = tracks_real[tracks_real[col] < np.inf][col].max()
+
+    initial, spac = get_diagram_arr_linspace(tracks_real, tracks_pred_true, start, end, n_ticks, col)
+    # mean line
+    # find number of ticks until no nans present
+
+    if style == 'boxplot':
+        interval_size = len(initial) / n_boxes
+        subarrays = {}
+        positions = {}
+        means = {}
+        stds = {}
+        first = 0
+        #get_boxes
+        for interval in range(n_boxes):
+            second = int(first + interval_size)
+            values_in_interval = initial[first:min(second, len(initial))]
+            pos_in_interval = spac[first:min(second, len(initial))]
+            means[interval] = np.mean(values_in_interval)
+            stds[interval] = np.std(values_in_interval)
+            positions[interval] = np.mean(pos_in_interval)
+            first = second
+        draw_from_data(title=f'{model_name} track efficiency vs {col_pretty} ({total_events} events)',
```

Comment:
what is `col_pretty`, what does it mean?

Reply:
code of this function is mostly written by Egor :)
col_pretty means  formatted math name of column, such as cos_x written as formula